db/migrate/20171012202650_cleanup_invalid_contacts.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+class CleanupInvalidContacts < ActiveRecord::Migration[5.1]
+  class Contact < ApplicationRecord
+    belongs_to :user
+    belongs_to :person
+
+    has_many :aspect_memberships, dependent: :destroy
+    has_many :aspects, through: :aspect_memberships
+
+    before_destroy :destroy_notifications
+
+    def destroy_notifications
+      Notification.where(
+        target_type:  "Person",
+        target_id:    person_id,
+        recipient_id: user_id,
+        type:         "Notifications::StartedSharing"
+      ).destroy_all
+    end
+  end
+
+  class User < ApplicationRecord
+    has_one :person, inverse_of: :owner, foreign_key: :owner_id
```

Comment:
Since you join from contacts via person to user, you should add the person model with the user relation. This line can be removed then, because it isn't needed.